### PR TITLE
Allow omitempty flag to be used with INSERT statements

### PR DESCRIPTION
**Pull request by Aflynn50**
When inserting a into a table with unique primary keys the user needs to specify the primary key in their object every time. If it is left to its default zero value it will most likely break the unique constraint. In some tables, the primary key can is auto when a new row is inserted. In this case the user may want to insert a full struct using the asterisk notation but omitting the primary key.

To automatically omit a primary key, the user can set a flag in the `db` tag after the column name for the field e.g. 
```
type Person struct {
	ID        int    `db:"id, omitempty"`
	Name      string `db:"name"`
	PostCode  int    `db:"address_id"`
}
```
This will indicate to SQLair that if this value is zero when the parameter is passed for execution, it should not be inserted into the database.

As specified in the [INSERT statement support specification](https://docs.google.com/document/d/1DMKPS85msymrdbA5-EoP4Sf6b9blSAjtj367EuEmP7Y/edit), the omit empty flag should only work for struct fields being inserted via an asterisk. If the field is inserted explicitly, e.g. `INSERT INTO t (id) VALUES ($Person.id)` then we should throw an error. Similarly, if a tag with `omitempty` set is used in an input expression elsewhere, an error should also be thrown e.g. `SELECT &Person.* FROM person WHERE id = $Person.id`.

### CODE REVIEW

File: internal/expr/bindinputs.go [hunk 1]
```diff
@@ -41,66 +41,61 @@ func (tbe *TypeBoundExpr) BindInputs(args ...any) (pq *PrimedQuery, err error) {
...
 					return nil, err
 				}
-				argTypeUsed[input.ArgType()] = true
-				for j, val := range vals {
-					if j != 0 {
-						sqlStr.WriteString(", ")
+				if len(vals) > 1 {
+					// Only slices return multiple values and they are not
+					// allowed in insert expressions.
+					return nil, fmt.Errorf("internal error: unexpected values")
```
Comment (manadart) [inline internal/expr/bindinputs.go]: The error might be more descriptive if the text is along the lines of the comment above.

File: internal/expr/bindinputs.go [hunk 2]
```diff
@@ -142,11 +132,20 @@ type typedInputExpr struct {
...
+	// query. If the column is inserted via an asterisk type, it is false.
+	explicit bool
+}
+
 // typedInsertExpr stores information about the Go values to use as inputs inside
 // an INSERT statement.
 type typedInsertExpr struct {
-	columns []string
-	inputs  []typeinfo.Input
+	insertColumns []insertColumn
```
Comment (manadart) [inline internal/expr/bindinputs.go:157]: I like the way you're thinking here.

File: internal/expr/bindinputs.go [hunk 3]
```diff
@@ -155,6 +154,61 @@ type typedOutputExpr struct {
 	outputColumns []outputColumn
 }
 
+type sqlBuilder struct {
```
Comment (manadart) [inline internal/expr/bindinputs.go:168]: This encapsulation is a nice addition.

File: internal/expr/bindinputs.go [hunk 4]
```diff
@@ -155,6 +154,61 @@ type typedOutputExpr struct {
...
+	buf bytes.Buffer
+}
+
+// writeInsert writes the SQL for INSERT statements to the sqlBuilder.
+func (b *sqlBuilder) writeInsert(inputCount int, columns []string) {
+	// Write out the columns.
+	b.buf.WriteString("(")
+	for i, col := range columns {
+		if i != 0 {
```
Comment (manadart) [inline internal/expr/bindinputs.go]: This repeated block could be a method accepting `i` and the string to write (`col` in this case).

File: internal/expr/bindtypes.go [hunk 5]
```diff
@@ -142,26 +142,33 @@ func (e *asteriskInsertExpr) bindTypes(argInfo typeinfo.ArgInfo) (tie any, err e
...
 		if source.memberName == "*" {
-			inps, tags, err := argInfo.AllStructInputs(source.typeName)
+			inputs, tags, err := argInfo.AllStructInputs(source.typeName)
 			if err != nil {
 				return nil, err
 			}
-			columns = append(columns, tags...)
-			inputs = append(inputs, inps...)
+			for i, input := range inputs {
+				cols = append(cols, insertColumn{
```
Comment (manadart) [inline internal/expr/bindtypes.go]: Worth a constructor that makes this `insertColumn`, then these are nice one-line functions.

**Review by manadart — APPROVED** [review]
Nice job on this one. A little house-keeping and we're good.

File: internal/expr/bindinputs.go (hunk 1 above)
Comment (Aflynn50) [reply internal/expr/bindinputs.go]: Changed.

File: internal/expr/bindinputs.go (hunk 4 above)
Comment (Aflynn50) [reply internal/expr/bindinputs.go]: Added a `sqlBuilder.writeCommaSeperatedList` which accepts a writer function for printing each list element to the buffer.

File: internal/expr/bindtypes.go (hunk 5 above)
Comment (Aflynn50) [reply internal/expr/bindtypes.go]: Added the constructor. This now also mirrors the `newOutputColumn` constructor.

**Pull request merged**